locations/storefinder_detector.py

```diff
@@ -0,0 +1,447 @@
+        self,
+        url: str = None,
+        brand_wikidata: str = None,
+        operator_wikidata: str = None,
+        spider_key: str = None,
+        spider_class_name: str = None,
+        *args,
+        **kwargs,
+    ):
+        self.start_urls = [url]
+        self.brand_wikidata = brand_wikidata
+        self.operator_wikidata = operator_wikidata
+        self.spider_key = spider_key
+        self.spider_class_name = spider_class_name
+        self.automatically_set_parameters()
+
+    def automatically_set_brand_or_operator_from_start_url(self):
+        """
+        Automatically extract parameters["brand_wikidata"] or
+        parameters["operator_wikidata"] from a supplied
+        start_urls[0]. Name Suggestion Index data is searched to find
+        an entry that has the same domain name as start_urls[0]. If
+        a match is found, parameters["brand_wikidata"] or
+        parameters["operator_wikidata"] are automatically set.
+        """
+        if not self.parameters["brand_wikidata"] and not self.parameters["operator_wikidata"] and self.start_urls:
+            nsi = NSI()
+            wikidata_code = nsi.get_wikidata_code_from_url(self.start_urls[0])
+            if wikidata_code:
+                nsi_matches = [nsi_match for nsi_match in nsi.iter_nsi(wikidata_code)]
+                if len(nsi_matches) == 1:
+                    if "brand:wikidata" in nsi_matches[0]["tags"].keys():
+                        self.parameters["brand_wikidata"] = wikidata_code
+                    elif "operator:wikidata" in nsi_matches[0]["tags"].keys():
+                        self.parameters["operator_wikidata"] = wikidata_code
+
+    def automatically_set_parameters(self):
+        """
+        Automatically extract parameters from at least one or more
+        of the following:
+          1. start_urls[0]
+          2. parameters["brand_wikidata"]
+          3. parameters["operator_wikidata"]
+
+        If any of the above are specified for this spider, and this
+        method is called, this method will attempt to populate all
+        other parameters automatically from Name Suggestion Index
+        data.
+
+        See automatically_set_brand_or_operator_from_start_url(self)
+        for details on how parameters are extracted from just a single
+        supplied start_urls[0].
+        """
+        if not self.parameters["brand_wikidata"] and not self.parameters["operator_wikidata"]:
+            self.automatically_set_brand_or_operator_from_start_url()
+
+        nsi = NSI()
+        if self.parameters["brand_wikidata"] or self.parameters["operator_wikidata"]:
+            wikidata_code = self.parameters["brand_wikidata"] or self.parameters["operator_wikidata"]
+            nsi_matches = [nsi_match for nsi_match in nsi.iter_nsi(wikidata_code)]
+            if len(nsi_matches) != 1:
+                return
+            spider_key = re.sub(r"[^a-zA-Z0-9_]", "", nsi_matches[0]["tags"]["name"].replace(" ", "_")).lower()
```

Comment:
Change request:
Sometimes the generated names end up looking like:
Go! Games & Toys => go_games__toys_us

Can the name generation:

* remove all repeated _
* detect if it starts with a _ and remove that
* detect if it is only the country code remaining and fallback to something like the wikidata id as the spider name or something else?